pkg/volume/nfs/nfs.go

```diff
@@ -84,105 +77,46 @@ func (plugin *nfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cl
 
 func (plugin *nfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter nfsMountInterface) (volume.Cleaner, error) {
 	return &nfs{
-		volName:    volName,
-		server:     "",
-		exportPath: "",
-		readOnly:   false,
-		mounter:    mounter,
-		podRef:     &api.ObjectReference{UID: podUID},
-		plugin:     plugin,
+		NetworkVolume: &network_volume.NetworkVolume{
+			VolName: volName,
+			Plugin:  plugin,
+			PodUID:  podUID,
+			Mounter: &nfsNetworkVolumeMounter{
+				server:     "",
+				exportPath: "",
+				readOnly:   false,
+				mounter:    mounter,
+			},
+		},
 	}, nil
 }
 
 // NFS volumes represent a bare host file or directory mount of an NFS export.
 type nfs struct {
-	volName    string
-	podRef     *api.ObjectReference
+	*network_volume.NetworkVolume
+}
+
+type nfsNetworkVolumeMounter struct {
 	server     string
 	exportPath string
 	readOnly   bool
 	mounter    nfsMountInterface
-	plugin     *nfsPlugin
-}
-
-// SetUp attaches the disk and bind mounts to the volume path.
-func (nfsVolume *nfs) SetUp() error {
-	return nfsVolume.SetUpAt(nfsVolume.GetPath())
 }
 
-func (nfsVolume *nfs) SetUpAt(dir string) error {
-	mountpoint, err := nfsVolume.mounter.IsMountPoint(dir)
-	glog.V(4).Infof("NFS mount set up: %s %v %v", dir, mountpoint, err)
-	if err != nil && !os.IsNotExist(err) {
-		return err
-	}
-	if mountpoint {
-		return nil
-	}
-	exportDir := nfsVolume.exportPath
-	os.MkdirAll(dir, 0750)
-	err = nfsVolume.mounter.Mount(nfsVolume.server, exportDir, dir, nfsVolume.readOnly)
-	if err != nil {
-		mountpoint, mntErr := nfsVolume.mounter.IsMountPoint(dir)
-		if mntErr != nil {
-			glog.Errorf("IsMountpoint check failed: %v", mntErr)
-			return err
-		}
-		if mountpoint {
-			if mntErr = nfsVolume.mounter.Unmount(dir); mntErr != nil {
-				glog.Errorf("Failed to unmount: %v", mntErr)
-				return err
-			}
-			mountpoint, mntErr := nfsVolume.mounter.IsMountPoint(dir)
-			if mntErr != nil {
-				glog.Errorf("IsMountpoint check failed: %v", mntErr)
-				return err
-			}
-			if mountpoint {
-				// This is very odd, we don't expect it.  We'll try again next sync loop.
-				glog.Errorf("%s is still mounted, despite call to unmount().  Will try again next sync loop.", dir)
-				return err
-			}
-		}
-		os.Remove(dir)
-		return err
-	}
-	return nil
+func (mounter *nfsNetworkVolumeMounter) Mount(dest string) error {
+	exportDir := mounter.exportPath
+	return mounter.mounter.Mount(mounter.server, exportDir, dest, mounter.readOnly)
```

Comment:
Yeah I agree... I made NetworkVolumeMounter extend mount.Interface. That takes care of the mounter.mounter pattern, and eliminates the need for call forwarding 'IsMountPoint' function needed for testing.